File: go/core/ecdh.go
package core

import (
	"crypto/elliptic"
	"crypto/rand"
	"errors"
	"math/big"
)

type ECDH struct {
	priv []byte
	x    *big.Int
	y    *big.Int
}

func NewECDH() (*ECDH, error) {
	priv, x, y, err := elliptic.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		return nil, err
	}
	return &ECDH{priv: priv, x: x, y: y}, nil
}

// NewECDHFromPrivate rebuilds the keypair from a 32-byte P-256 private scalar,
// so the server's identity key can be persisted across restarts.
func NewECDHFromPrivate(priv []byte) (*ECDH, error) {
	if len(priv) != 32 {
		return nil, errors.New("invalid private key length")
	}
	x, y := elliptic.P256().ScalarBaseMult(priv)
	if x == nil || y == nil {
		return nil, errors.New("invalid private key")
	}
	return &ECDH{priv: priv, x: x, y: y}, nil
}

func (e *ECDH) GetPrivateKey() []byte {
	return e.priv
}

func (e *ECDH) GetPublicKey() []byte {
	return elliptic.Marshal(elliptic.P256(), e.x, e.y)
}

func (e *ECDH) ComputeSecret(peerPubKey []byte) ([]byte, error) {
	px, py := elliptic.Unmarshal(elliptic.P256(), peerPubKey)
	if px == nil || py == nil {
		return nil, errors.New("invalid peer public key")
	}

	zx, _ := elliptic.P256().ScalarMult(px, py, e.priv)
	
	// Ensure the secret is always 32 bytes (pad with leading zeros if necessary)
	secret := zx.Bytes()
	if len(secret) < 32 {
		padded := make([]byte, 32)
		copy(padded[32-len(secret):], secret)
		return padded, nil
	}
	return secret[:32], nil
}

File: rust/src/core/ecdh.rs
use p256::{
    ecdh::diffie_hellman,
    PublicKey, SecretKey,
};
use sha2::{Digest, Sha256};

#[derive(Clone)]
pub struct ECDH {
    secret: SecretKey,
    public_key: PublicKey,
}

impl ECDH {
    pub fn new() -> Self {
        let secret = SecretKey::random(&mut rand::rng());
        let public_key = secret.public_key();
        Self { secret, public_key }
    }

    // Rebuilds the keypair from a 32-byte P-256 private scalar, so the server's
    // identity key can be persisted across restarts.
    pub fn from_private_key(bytes: &[u8]) -> Result<Self, String> {
        let secret = SecretKey::from_slice(bytes).map_err(|e| format!("invalid private key: {:?}", e))?;
        let public_key = secret.public_key();
        Ok(Self { secret, public_key })
    }

    pub fn private_key_bytes(&self) -> [u8; 32] {
        self.secret.to_bytes().into()
    }

    pub fn get_public_key(&self) -> Vec<u8> {
        self.public_key.to_sec1_bytes().to_vec()
    }

    pub fn compute_secret(&self, peer_pub_key_bytes: &[u8]) -> Result<Vec<u8>, String> {
        let peer_pub = PublicKey::from_sec1_bytes(peer_pub_key_bytes).map_err(|_| "Invalid peer public key")?;
        let shared = diffie_hellman(self.secret.to_nonzero_scalar(), peer_pub.as_affine());
        let secret_bytes = shared.raw_secret_bytes();
        Ok(secret_bytes.to_vec())
    }

    // Derives the per-sender MAC key shared with a peer: SHA-256(ECDH(priv,
    // peerPub)). Both sides compute the same value from their own private key,
    // so a third member cannot derive it and cannot forge the other member's
    // control frames.
    pub fn pairwise_key(&self, peer_pub_key_bytes: &[u8]) -> Result<Vec<u8>, String> {
        let secret = self.compute_secret(peer_pub_key_bytes)?;
        Ok(Sha256::digest(&secret).to_vec())
    }
}
